lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+    # require expand does alias, updates context_modules and runtime_modules
+    # we do it here instead
+
+    env = %{env | context_modules: [full | env.context_modules]}
+
+    state =
+      case original_env do
+        %{function: nil} ->
+          state
+
+        _ ->
+          %{state | runtime_modules: [full | state.runtime_modules]}
+      end
+
+    range = extract_range(meta)
+
+    module_functions =
+      case state.protocol do
+        nil -> []
+        _ -> [{:__impl__, [:atom], :def}]
+      end
+
+    state =
+      state
+      |> add_module_to_index(full, range, [])
+      |> add_module
+      |> add_current_env_to_line(meta, %{env | module: full})
+      |> add_module_functions(%{env | module: full}, module_functions, range)
+      |> new_vars_scope
+      |> new_attributes_scope
+
+    {state, _env} = maybe_add_protocol_behaviour(state, %{env | module: full})
+
+    {_result, state, e_env} = expand(block, state, %{env | module: full})
+
+    # here we handle module callbacks. Only before_compile macro callbacks are expanded as they
+    # affect module body. Func before_compile callbacks are not executed. after_compile and after_verify
+    # are not executed as we do not preform a real compilation
+    {state, _e_env} =
+      for args <- Map.get(state.attribute_store, {full, :before_compile}, []) do
+        case args do
+          {module, fun} -> [module, fun]
+          module -> [module, :__before_compile__]
+        end
+      end
+      |> Enum.reduce({state, e_env}, fn target, {state, env} ->
+        # module vars are not accessible in module callbacks
+        env = %{env | versioned_vars: %{}, line: meta[:line]}
+        state_orig = state
+        state = new_func_vars_scope(state)
+
+        # elixir dispatches callbacks by raw dispatch and eval_forms
+        # instead we expand a bock with require and possibly expand macros
+        # we do not attempt to exec function callbacks
+        ast =
+          {:__block__, [],
+           [
+             {:require, [], [hd(target)]},
+             {{:., [], target}, [], [env]}
+           ]}
+
+        {_result, state, env} = expand(ast, state, env)
+        {remove_func_vars_scope(state, state_orig), env}
+      end)
+
+    # restore vars from outer scope
+    # restore version counter
+    state =
+      state
+      |> apply_optional_callbacks(%{env | module: full})
+      |> remove_vars_scope(state_orig, true)
+      |> remove_attributes_scope
+      |> remove_module
+
+    # in elixir the result of defmodule expansion is
+    # require (a module atom) and :elixir_module.compile dot call in block
+    # we don't need that
+
+    {{:__block__, [], []}, state, env}
+  end
+
+  defp expand_macro(
+         meta,
+         Protocol,
+         :def,
+         [{name, _, _args = [_ | _]} = call],
+         callback,
+         state,
+         env
+       )
+       when is_atom(name) do
+    # transform protocol def to def with empty body
```

Comment:
Wouldn't the regular macro expansion do this anyway? Why do we need to intercept?

Reply:
Eval. I also collect some info about protocols and implementations